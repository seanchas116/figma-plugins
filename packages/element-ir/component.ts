export interface Component {
  key?: string;
  element: Element;
  propertyDefinitions: PropertyDefinition[];
}

export interface PropertyDefinition {
  key: string;
  type: PropertyType;
}

export type PropertyType =
  | {
      type: "boolean";
    }
  | {
      type: "string";
    }
  | {
      type: "enum";
      options: string[];
    }
  | {
      type: "instance";
    };
