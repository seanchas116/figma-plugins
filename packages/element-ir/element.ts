import {
  FrameStyle,
  ImageStyle,
  InstanceStyle,
  SVGStyle,
  TextStyle,
} from "./style";

export interface CommonProps {
  id: string;
  name: string;
  propertyRef: {
    visible?: string;
    content?: string;
    component?: string;
  };
}

export interface FrameElement extends CommonProps {
  type: "frame";
  style: FrameStyle;
  children: Element[];
}

export interface TextElement extends CommonProps {
  type: "text";
  style: TextStyle;
  content: string; // TODO: spans
}

export interface ImageElement extends CommonProps {
  type: "image";
  style: ImageStyle;
  imageID: string;
}

export interface SVGElement extends CommonProps {
  type: "svg";
  style: SVGStyle;
  svg: string;
}

export interface InstanceElement extends CommonProps {
  type: "instance";
  componentKey: string;
  properties: Record<string, any>;
  // overrides are not supported (instances are inlined if any overrides are present)
  style: InstanceStyle;
}

export type Element =
  | FrameElement
  | TextElement
  | ImageElement
  | SVGElement
  | InstanceElement;
