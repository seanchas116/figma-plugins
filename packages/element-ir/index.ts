export * from "./style";
export * from "./element";
export * from "./component";
