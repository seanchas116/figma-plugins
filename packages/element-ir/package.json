{
  "name": "@uimix/element-ir",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "typings": "index.d.ts",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "MIT"
}
