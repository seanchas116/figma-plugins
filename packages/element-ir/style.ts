export interface Color {
  r: number;
  g: number;
  b: number;
  a: number;
}

export interface SolidPaint {
  type: "solid";
  color: Color;
}

export interface Vector {
  x: number;
  y: number;
}

export interface ColorStop {
  position: number;
  color: Color;
}

export interface GradientPaint {
  type: "gradientLinear" | "gradientRadial" | "gradientAngular";
  handles: [Vector, Vector, Vector];
  stops: ColorStop[];
}

export interface ImagePaint {
  type: "image";
  imageID: string;
  size: "fill" | "contain" | "cover";
}

export type Paint = SolidPaint | GradientPaint | ImagePaint;

export interface DimensionStyleMixin {
  position: "absolute" | "relative";
  display: "flex" | "none";
  x:
    | { left: number }
    | { right: number }
    | { centerRatio: number }
    | { left: number; right: number };
  y:
    | { top: number }
    | { bottom: number }
    | { centerRatio: number }
    | { top: number; bottom: number };
  width: number | "fit-content" | "stretch";
  height: number | "fit-content" | "stretch";
}

export interface RectangleFillBorderStyleMixin {
  borderTopLeftRadius: number;
  borderTopRightRadius: number;
  borderBottomLeftRadius: number;
  borderBottomRightRadius: number;

  border: Paint[];
  borderTopWidth: number;
  borderRightWidth: number;
  borderBottomWidth: number;
  borderLeftWidth: number;

  background: Paint[];
}

export interface FrameStyleMixin {
  overflow: "hidden" | "visible"; // TODO: scroll
  flexDirection: "row" | "column";
  gap: number;
  paddingTop: number;
  paddingBottom: number;
  paddingLeft: number;
  paddingRight: number;
  alignItems: "center" | "flex-start" | "flex-end" | "baseline";
  justifyContent: "center" | "flex-start" | "flex-end" | "space-between";
}

export interface TextSpanStyleMixin {
  fontFamily: string;
  fontStyle: "normal" | "italic";
  fontWeight: number;
  fontSize: number;
  lineHeight: number | "normal";
  letterSpacing: number; // em;
  color: Paint[];
}

export interface TextStyleMixin {
  textAlign: "left" | "center" | "right" | "justify";
  justifyContent: "center" | "flex-start" | "flex-end"; // assuming a text element becomes a vertical flexbox
}

export interface ImageStyleMixin {
  objectFit: "fill" | "contain" | "cover";
}

export interface FrameStyle
  extends DimensionStyleMixin,
    RectangleFillBorderStyleMixin,
    FrameStyleMixin {}

export interface TextStyle
  extends DimensionStyleMixin,
    TextSpanStyleMixin,
    TextStyleMixin {}

export interface SVGStyle extends DimensionStyleMixin {}

export interface ImageStyle
  extends DimensionStyleMixin,
    RectangleFillBorderStyleMixin,
    ImageStyleMixin {}
