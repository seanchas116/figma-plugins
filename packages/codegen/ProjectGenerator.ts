import { Component } from "@uimix/element-ir";
import { camelCase, capitalize } from "lodash-es";
import {
  ExtendedComponent,
  ExtendedPropertyDefinition,
  GeneratedFile,
} from "./types";
import { ComponentGenerator } from "./ComponentGenerator";
import { IDGenerator } from "./util/IDGenerator";
import { Config } from "./Config";

export interface ProjectGeneratorOptions {
  config: Config;
  components: Component[];
}

export class ProjectGenerator {
  constructor(options: ProjectGeneratorOptions) {
    const componentIDGenerator = new IDGenerator();

    for (const component of options.components) {
      const propertyNameGenerator = new IDGenerator();
      const propertyForName = new Map<string, ExtendedPropertyDefinition>();

      for (const prop of component.propertyDefinitions) {
        const inCodeName = propertyNameGenerator.generate(
          camelCase(prop.name.split("#")[0])
        );
        propertyForName.set(prop.name, {
          ...prop,
          inCodeName,
        });
      }

      const extendedComponent: ExtendedComponent = {
        ...component,
        inCodeName: componentIDGenerator.generate(
          capitalize(camelCase(component.element.name ?? ""))
        ),
        propertyForName,
      };

      if (component.key) {
        console.log("component.key", component.key);
        this.components.set(component.key, extendedComponent);
      }

      const componentGenerator = new ComponentGenerator(extendedComponent, {
        config: options.config,
        otherComponents: this.components,
      });
      this.componentGenerators.push(componentGenerator);
    }
  }

  readonly componentGenerators: ComponentGenerator[] = [];
  readonly components = new Map<string, ExtendedComponent>();

  generate(): GeneratedFile[] {
    return this.componentGenerators.flatMap((c) => c.generate());
  }
}
