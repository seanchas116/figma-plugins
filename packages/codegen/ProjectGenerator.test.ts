import { describe, expect, it } from "vitest";
import { Component } from "@uimix/element-ir";

import _components from "./__fixtures__/components.json";
import { ProjectGenerator } from "./ProjectGenerator";
const components: Component[] = _components as any;

describe("Generator", () => {
  for (const style of ["tailwind", "inline", "css"] as const) {
    it(`should emit ${style}`, () => {
      const generator = new ProjectGenerator({
        components,
        style,
      });

      expect(generator.generate()).toMatchSnapshot();
    });
  }
});
