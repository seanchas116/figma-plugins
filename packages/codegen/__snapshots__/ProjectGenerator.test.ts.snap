// Vitest Snapshot v1

exports[`Generator > should emit css 1`] = `
[
  {
    "content": ".Button {
  display: flex;
  position: relative;
  width: fit-content;
  height: fit-content;
  background: #3b82f6;
  border-radius: 4px 4px 4px 4px;
  overflow: hidden;
  gap: 6px;
  padding: 6px 12px 6px 12px;
  align-items: center;
  justify-content: center;
}
.Button__Button {
  display: flex;
  position: relative;
  width: fit-content;
  height: fit-content;
  font-family: Inter, sans-serif;
  font-size: 12px;
  font-weight: 500;
  color: #ffffff;
  text-align: left;
  flex-direction: column;
}
.Button__chevron {
  display: flex;
  position: relative;
  width: 8px;
  height: 8px;
  align-items: flex-start;
}
.Button__arrow_forward_ios {
  display: flex;
  position: absolute;
  left: 2.0833332538604736px;
  top: 0.6666666865348816px;
  width: 3.924999952316284px;
  height: 6.666666507720947px;
}
",
    "filePath": "Button.css",
  },
  {
    "content": "import \\"./Button.css\\";

export function Button(props) {
  return (
    <div {...props} className={\\"Button \\" + props.className}>
      <div className=\\"Button__Button\\">{props.text}</div>
      {props.showChevron && (
        <div className=\\"Button__chevron\\">
          <svg
            width={5}
            height={7}
            viewBox={\\"0 0 5 7\\"}
            fill={\\"none\\"}
            className=\\"Button__arrow_forward_ios\\"
          >
            <path
              d=\\"M0.674995 6.83334L0.0833282 6.24167L2.825 3.5L0.0833282 0.758338L0.674995 0.166672L4.00833 3.5L0.674995 6.83334Z\\"
              fill=\\"white\\"
            />
          </svg>
        </div>
      )}
    </div>
  );
}
",
    "filePath": "Button.js",
  },
  {
    "content": ".Form2 {
  display: flex;
  position: relative;
  width: fit-content;
  height: fit-content;
  background: #ffffff;
  overflow: hidden;
  flex-direction: column;
  gap: 13px;
  padding: 16px 16px 16px 16px;
  align-items: flex-start;
}
.Form2__h1_____ {
  display: flex;
  position: relative;
  width: fit-content;
  height: fit-content;
  font-family: Inter, sans-serif;
  font-size: 12px;
  font-weight: 400;
  color: #000000;
  text-align: left;
  flex-direction: column;
}
.Form2__Button {
  align-self: stretch;
  width: auto;
}
",
    "filePath": "Form2.css",
  },
  {
    "content": "import { Button } from \\"./Button.js\\";
import \\"./Form2.css\\";

export function Form2(props) {
  return (
    <div {...props} className={\\"Form2 \\" + props.className}>
      <div className=\\"Form2__h1_____\\">Some Form (stretch)</div>
      <Button showChevron={false} text={\\"OK\\"} className=\\"Form2__Button\\" />
    </div>
  );
}
",
    "filePath": "Form2.js",
  },
  {
    "content": ".Form {
  display: flex;
  position: relative;
  width: fit-content;
  height: fit-content;
  background: #ffffff;
  overflow: hidden;
  flex-direction: column;
  gap: 13px;
  padding: 16px 16px 16px 16px;
  align-items: flex-start;
}
.Form__h1_____ {
  display: flex;
  position: relative;
  width: fit-content;
  height: fit-content;
  font-family: Inter, sans-serif;
  font-size: 12px;
  font-weight: 400;
  color: #000000;
  text-align: left;
  flex-direction: column;
}
.Form__Button {
}
",
    "filePath": "Form.css",
  },
  {
    "content": "import { Button } from \\"./Button.js\\";
import \\"./Form.css\\";

export function Form(props) {
  return (
    <div {...props} className={\\"Form \\" + props.className}>
      <div className=\\"Form__h1_____\\">Some Form</div>
      <Button showChevron={true} text={\\"OK\\"} className=\\"Form__Button\\" />
    </div>
  );
}
",
    "filePath": "Form.js",
  },
  {
    "content": ".Main {
  display: flex;
  position: relative;
  width: 353px;
  height: 205px;
  background: #ffffff;
  overflow: hidden;
  flex-direction: column;
  gap: 12px;
  padding: 16px 16px 16px 16px;
  align-items: flex-start;
}
.Main__Hello__UIMix_ {
  display: flex;
  position: relative;
  width: fit-content;
  height: fit-content;
  font-family: Inter, sans-serif;
  font-size: 16px;
  font-weight: 700;
  line-height: 1.5;
  color: #000000;
  text-align: left;
  flex-direction: column;
}
.Main__Lorem_ipsum_dolor_sit_amet__consectetur_adipiscing_elit__sed_do_eiusmod_tempor_incididunt_ut_labore_et_dolore_magna_aliqua_ {
  display: flex;
  position: relative;
  align-self: stretch;
  height: fit-content;
  font-family: Inter, sans-serif;
  font-size: 16px;
  font-weight: 400;
  line-height: 1.5;
  color: #000000;
  text-align: left;
  flex-direction: column;
}
.Main__Button {
  align-self: stretch;
  width: auto;
}
",
    "filePath": "Main.css",
  },
  {
    "content": "import { Button } from \\"./Button.js\\";
import \\"./Main.css\\";

export function Main(props) {
  return (
    <div {...props} className={\\"Main \\" + props.className}>
      <div className=\\"Main__Hello__UIMix_\\">Hello, UIMix!</div>
      <div className=\\"Main__Lorem_ipsum_dolor_sit_amet__consectetur_adipiscing_elit__sed_do_eiusmod_tempor_incididunt_ut_labore_et_dolore_magna_aliqua_\\">
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua.
      </div>
      <Button showChevron={false} text={\\"OK\\"} className=\\"Main__Button\\" />
    </div>
  );
}
",
    "filePath": "Main.js",
  },
  {
    "content": ".Frame2 {
  display: flex;
  position: relative;
  width: 517px;
  height: fit-content;
  background: #ffffff;
  overflow: hidden;
  flex-direction: column;
  align-items: flex-start;
}
.Frame2__Header {
  display: flex;
  position: relative;
  flex-grow: 0;
  align-self: stretch;
  width: auto;
  height: fit-content;
}
.Frame2__Frame_1 {
  display: flex;
  position: relative;
  align-self: stretch;
  height: fit-content;
  flex-direction: column;
  gap: 24px;
  padding: 24px 24px 24px 24px;
  align-items: flex-start;
}
.Frame2__Hello__UIMix_ {
  display: flex;
  position: relative;
  align-self: stretch;
  height: fit-content;
  font-family: Inter, sans-serif;
  font-size: 16px;
  font-weight: 700;
  line-height: 1.5;
  color: #000000;
  text-align: left;
  flex-direction: column;
}
.Frame2__Lorem_ipsum_dolor_sit_amet__consectetur_adipiscing_elit__sed_do_eiusmod_tempor_incididunt_ut_labore_et_dolore_magna_aliqua_ {
  display: flex;
  position: relative;
  align-self: stretch;
  height: fit-content;
  font-family: Inter, sans-serif;
  font-size: 16px;
  font-weight: 400;
  line-height: 1.5;
  color: #000000;
  text-align: left;
  flex-direction: column;
}
.Frame2__Button {
  display: flex;
  position: relative;
  flex-grow: 0;
  align-self: auto;
  width: fit-content;
  height: fit-content;
}
",
    "filePath": "Frame2.css",
  },
  {
    "content": "import \\"./Frame2.css\\";

export function Frame2(props) {
  return (
    <div {...props} className={\\"Frame2 \\" + props.className}>
      <Header productName={\\"Hello, world!\\"} className=\\"Frame2__Header\\" />
      <div className=\\"Frame2__Frame_1\\">
        <div className=\\"Frame2__Hello__UIMix_\\">Hello, UIMix!</div>
        <div className=\\"Frame2__Lorem_ipsum_dolor_sit_amet__consectetur_adipiscing_elit__sed_do_eiusmod_tempor_incididunt_ut_labore_et_dolore_magna_aliqua_\\">
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
          eiusmod tempor incididunt ut labore et dolore magna aliqua.
        </div>
        <Button
          primary={true}
          size={\\"large\\"}
          label={\\"Press Me\\"}
          backgroundColor={\\"\\"}
          className=\\"Frame2__Button\\"
        />
      </div>
    </div>
  );
}
",
    "filePath": "Frame2.js",
  },
]
`;

exports[`Generator > should emit inline 1`] = `
[
  {
    "content": "export function Button(props) {
  return (
    <div
      {...props}
      style={{
        display: \\"flex\\",
        position: \\"relative\\",
        width: \\"fit-content\\",
        height: \\"fit-content\\",
        background: \\"#3b82f6\\",
        borderRadius: \\"4px 4px 4px 4px\\",
        overflow: \\"hidden\\",
        gap: \\"6px\\",
        padding: \\"6px 12px 6px 12px\\",
        alignItems: \\"center\\",
        justifyContent: \\"center\\",
        ...props.style,
      }}
    >
      <div
        style={{
          display: \\"flex\\",
          position: \\"relative\\",
          width: \\"fit-content\\",
          height: \\"fit-content\\",
          fontFamily: \\"Inter, sans-serif\\",
          fontSize: \\"12px\\",
          fontWeight: 500,
          color: \\"#ffffff\\",
          textAlign: \\"left\\",
          flexDirection: \\"column\\",
        }}
      >
        {props.text}
      </div>
      {props.showChevron && (
        <div
          style={{
            display: \\"flex\\",
            position: \\"relative\\",
            width: \\"8px\\",
            height: \\"8px\\",
            alignItems: \\"flex-start\\",
          }}
        >
          <svg
            width={5}
            height={7}
            viewBox={\\"0 0 5 7\\"}
            fill={\\"none\\"}
            style={{
              display: \\"flex\\",
              position: \\"absolute\\",
              left: \\"2.0833332538604736px\\",
              top: \\"0.6666666865348816px\\",
              width: \\"3.924999952316284px\\",
              height: \\"6.666666507720947px\\",
            }}
          >
            <path
              d=\\"M0.674995 6.83334L0.0833282 6.24167L2.825 3.5L0.0833282 0.758338L0.674995 0.166672L4.00833 3.5L0.674995 6.83334Z\\"
              fill=\\"white\\"
            />
          </svg>
        </div>
      )}
    </div>
  );
}
",
    "filePath": "Button.js",
  },
  {
    "content": "import { Button } from \\"./Button.js\\";

export function Form2(props) {
  return (
    <div
      {...props}
      style={{
        display: \\"flex\\",
        position: \\"relative\\",
        width: \\"fit-content\\",
        height: \\"fit-content\\",
        background: \\"#ffffff\\",
        overflow: \\"hidden\\",
        flexDirection: \\"column\\",
        gap: \\"13px\\",
        padding: \\"16px 16px 16px 16px\\",
        alignItems: \\"flex-start\\",
        ...props.style,
      }}
    >
      <div
        style={{
          display: \\"flex\\",
          position: \\"relative\\",
          width: \\"fit-content\\",
          height: \\"fit-content\\",
          fontFamily: \\"Inter, sans-serif\\",
          fontSize: \\"12px\\",
          fontWeight: 400,
          color: \\"#000000\\",
          textAlign: \\"left\\",
          flexDirection: \\"column\\",
        }}
      >
        Some Form (stretch)
      </div>
      <Button
        showChevron={false}
        text={\\"OK\\"}
        style={{ alignSelf: \\"stretch\\", width: \\"auto\\" }}
      />
    </div>
  );
}
",
    "filePath": "Form2.js",
  },
  {
    "content": "import { Button } from \\"./Button.js\\";

export function Form(props) {
  return (
    <div
      {...props}
      style={{
        display: \\"flex\\",
        position: \\"relative\\",
        width: \\"fit-content\\",
        height: \\"fit-content\\",
        background: \\"#ffffff\\",
        overflow: \\"hidden\\",
        flexDirection: \\"column\\",
        gap: \\"13px\\",
        padding: \\"16px 16px 16px 16px\\",
        alignItems: \\"flex-start\\",
        ...props.style,
      }}
    >
      <div
        style={{
          display: \\"flex\\",
          position: \\"relative\\",
          width: \\"fit-content\\",
          height: \\"fit-content\\",
          fontFamily: \\"Inter, sans-serif\\",
          fontSize: \\"12px\\",
          fontWeight: 400,
          color: \\"#000000\\",
          textAlign: \\"left\\",
          flexDirection: \\"column\\",
        }}
      >
        Some Form
      </div>
      <Button showChevron={true} text={\\"OK\\"} style={{}} />
    </div>
  );
}
",
    "filePath": "Form.js",
  },
  {
    "content": "import { Button } from \\"./Button.js\\";

export function Main(props) {
  return (
    <div
      {...props}
      style={{
        display: \\"flex\\",
        position: \\"relative\\",
        width: \\"353px\\",
        height: \\"205px\\",
        background: \\"#ffffff\\",
        overflow: \\"hidden\\",
        flexDirection: \\"column\\",
        gap: \\"12px\\",
        padding: \\"16px 16px 16px 16px\\",
        alignItems: \\"flex-start\\",
        ...props.style,
      }}
    >
      <div
        style={{
          display: \\"flex\\",
          position: \\"relative\\",
          width: \\"fit-content\\",
          height: \\"fit-content\\",
          fontFamily: \\"Inter, sans-serif\\",
          fontSize: \\"16px\\",
          fontWeight: 700,
          lineHeight: 1.5,
          color: \\"#000000\\",
          textAlign: \\"left\\",
          flexDirection: \\"column\\",
        }}
      >
        Hello, UIMix!
      </div>
      <div
        style={{
          display: \\"flex\\",
          position: \\"relative\\",
          alignSelf: \\"stretch\\",
          height: \\"fit-content\\",
          fontFamily: \\"Inter, sans-serif\\",
          fontSize: \\"16px\\",
          fontWeight: 400,
          lineHeight: 1.5,
          color: \\"#000000\\",
          textAlign: \\"left\\",
          flexDirection: \\"column\\",
        }}
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua.
      </div>
      <Button
        showChevron={false}
        text={\\"OK\\"}
        style={{ alignSelf: \\"stretch\\", width: \\"auto\\" }}
      />
    </div>
  );
}
",
    "filePath": "Main.js",
  },
  {
    "content": "export function Frame2(props) {
  return (
    <div
      {...props}
      style={{
        display: \\"flex\\",
        position: \\"relative\\",
        width: \\"517px\\",
        height: \\"fit-content\\",
        background: \\"#ffffff\\",
        overflow: \\"hidden\\",
        flexDirection: \\"column\\",
        alignItems: \\"flex-start\\",
        ...props.style,
      }}
    >
      <Header
        productName={\\"Hello, world!\\"}
        style={{
          display: \\"flex\\",
          position: \\"relative\\",
          flexGrow: 0,
          alignSelf: \\"stretch\\",
          width: \\"auto\\",
          height: \\"fit-content\\",
        }}
      />
      <div
        style={{
          display: \\"flex\\",
          position: \\"relative\\",
          alignSelf: \\"stretch\\",
          height: \\"fit-content\\",
          flexDirection: \\"column\\",
          gap: \\"24px\\",
          padding: \\"24px 24px 24px 24px\\",
          alignItems: \\"flex-start\\",
        }}
      >
        <div
          style={{
            display: \\"flex\\",
            position: \\"relative\\",
            alignSelf: \\"stretch\\",
            height: \\"fit-content\\",
            fontFamily: \\"Inter, sans-serif\\",
            fontSize: \\"16px\\",
            fontWeight: 700,
            lineHeight: 1.5,
            color: \\"#000000\\",
            textAlign: \\"left\\",
            flexDirection: \\"column\\",
          }}
        >
          Hello, UIMix!
        </div>
        <div
          style={{
            display: \\"flex\\",
            position: \\"relative\\",
            alignSelf: \\"stretch\\",
            height: \\"fit-content\\",
            fontFamily: \\"Inter, sans-serif\\",
            fontSize: \\"16px\\",
            fontWeight: 400,
            lineHeight: 1.5,
            color: \\"#000000\\",
            textAlign: \\"left\\",
            flexDirection: \\"column\\",
          }}
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
          eiusmod tempor incididunt ut labore et dolore magna aliqua.
        </div>
        <Button
          primary={true}
          size={\\"large\\"}
          label={\\"Press Me\\"}
          backgroundColor={\\"\\"}
          style={{
            display: \\"flex\\",
            position: \\"relative\\",
            flexGrow: 0,
            alignSelf: \\"auto\\",
            width: \\"fit-content\\",
            height: \\"fit-content\\",
          }}
        />
      </div>
    </div>
  );
}
",
    "filePath": "Frame2.js",
  },
]
`;

exports[`Generator > should emit tailwind 1`] = `
[
  {
    "content": "import { twMerge } from \\"tailwind-merge\\";

export function Button(props) {
  return (
    <div
      {...props}
      className={twMerge(
        \\"flex relative w-fit bg-blue-500 rounded-tl rounded-tr rounded-br rounded-bl overflow-hidden gap-1.5 pt-1.5 pr-3 pb-1.5 pl-3 items-center justify-center\\",
        props.className
      )}
    >
      <div className=\\"w-fit font-['Inter',sans-serif] font-medium text-xs text-white\\">
        {props.text}
      </div>
      {props.showChevron && (
        <div className=\\"flex relative w-2 h-2 items-start\\">
          <svg
            width={5}
            height={7}
            viewBox={\\"0 0 5 7\\"}
            fill={\\"none\\"}
            className=\\"absolute left-[0.1302083283662796rem] top-[0.0416666679084301rem] w-[0.24531249701976776rem] h-[0.4166666567325592rem]\\"
          >
            <path
              d=\\"M0.674995 6.83334L0.0833282 6.24167L2.825 3.5L0.0833282 0.758338L0.674995 0.166672L4.00833 3.5L0.674995 6.83334Z\\"
              fill=\\"white\\"
            />
          </svg>
        </div>
      )}
    </div>
  );
}
",
    "filePath": "Button.js",
  },
  {
    "content": "import { Button } from \\"./Button.js\\";
import { twMerge } from \\"tailwind-merge\\";

export function Form2(props) {
  return (
    <div
      {...props}
      className={twMerge(
        \\"flex relative w-fit bg-white overflow-hidden flex-col gap-[0.8125rem] pt-4 pr-4 pb-4 pl-4 items-start\\",
        props.className
      )}
    >
      <div className=\\"w-fit font-['Inter',sans-serif] text-xs text-black\\">
        Some Form (stretch)
      </div>
      <Button showChevron={false} text={\\"OK\\"} className=\\"self-stretch w-auto\\" />
    </div>
  );
}
",
    "filePath": "Form2.js",
  },
  {
    "content": "import { Button } from \\"./Button.js\\";
import { twMerge } from \\"tailwind-merge\\";

export function Form(props) {
  return (
    <div
      {...props}
      className={twMerge(
        \\"flex relative w-fit bg-white overflow-hidden flex-col gap-[0.8125rem] pt-4 pr-4 pb-4 pl-4 items-start\\",
        props.className
      )}
    >
      <div className=\\"w-fit font-['Inter',sans-serif] text-xs text-black\\">
        Some Form
      </div>
      <Button showChevron={true} text={\\"OK\\"} className=\\"\\" />
    </div>
  );
}
",
    "filePath": "Form.js",
  },
  {
    "content": "import { Button } from \\"./Button.js\\";
import { twMerge } from \\"tailwind-merge\\";

export function Main(props) {
  return (
    <div
      {...props}
      className={twMerge(
        \\"flex relative w-[22.0625rem] h-[12.8125rem] bg-white overflow-hidden flex-col gap-3 pt-4 pr-4 pb-4 pl-4 items-start\\",
        props.className
      )}
    >
      <div className=\\"w-fit font-['Inter',sans-serif] font-bold text-base text-black\\">
        Hello, UIMix!
      </div>
      <div className=\\"self-stretch font-['Inter',sans-serif] text-base text-black\\">
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua.
      </div>
      <Button showChevron={false} text={\\"OK\\"} className=\\"self-stretch w-auto\\" />
    </div>
  );
}
",
    "filePath": "Main.js",
  },
  {
    "content": "import { twMerge } from \\"tailwind-merge\\";

export function Frame2(props) {
  return (
    <div
      {...props}
      className={twMerge(
        \\"flex relative w-[32.3125rem] bg-white overflow-hidden flex-col items-start\\",
        props.className
      )}
    >
      <Header
        productName={\\"Hello, world!\\"}
        className=\\"flex relative flex-none self-stretch w-auto h-fit\\"
      />
      <div className=\\"flex relative self-stretch flex-col gap-6 pt-6 pr-6 pb-6 pl-6 items-start\\">
        <div className=\\"self-stretch font-['Inter',sans-serif] font-bold text-base text-black\\">
          Hello, UIMix!
        </div>
        <div className=\\"self-stretch font-['Inter',sans-serif] text-base text-black\\">
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
          eiusmod tempor incididunt ut labore et dolore magna aliqua.
        </div>
        <Button
          primary={true}
          size={\\"large\\"}
          label={\\"Press Me\\"}
          backgroundColor={\\"\\"}
          className=\\"flex relative flex-none self-auto w-fit h-fit\\"
        />
      </div>
    </div>
  );
}
",
    "filePath": "Frame2.js",
  },
]
`;
